### PR TITLE
DRILL-6453: Resolve deadlock when reading from build and probe sides simultaneously in HashJoin

### PR DESCRIPTION
**Pull request by ilooner**
# The Problem

Originally hash join sniffed the first data holding batch in the probe and build side. Using the size statistics from both sides, memory calculations were performed in order to determine when to spill data. 

The issue with this is that fetching the first data holding batch from both sides can cause a deadlock in the exchange operators. Details of how this can happen have been included by others on the Jira.

# Theory of Operation

## Sniffing Batches

Batch sniffing is done in three phases.

 1. Schema sniffing is done in buildSchema()
 2. Before executing the build phase we sniff the first data holding build side batch and use the stats to decide the number of partitions and do memory calculations.
 3. Before executing the probe phase we sniff the first data holding probe side batch, and use the size statistics to do memory calculations that decide when to spill.

## Memory Estimation

When sniffing the schema for the build and probe side, we may get lucky and get data for the probe side. If this is the case then we cause use the probe side data to estimate the optimal number of partitions to use in the join operator. If we don't have probe side data when computing the number of partitions to use we assume that the incoming probe batches will be less than or equal to the configured batch size.

Since the number of partitions must be configured upfront before we may have probe data, we may get stuck in a situation where we have too many partitions to effectively process the probe side. In order to avoid this scenario we also adjust the number of records in probe side partition batches after we receive data from the probe side.

## Corner cases

While implementing this many corner cases had to be handled.

- Empty build side
- Empty probe side
- Empty probe and build sides
- Getting probe side data when retrieving the probe schema
- Not getting probe side data when retrieving the probe schema

## Testing

I added unit tests for all the corner cases, and have extracted logic for predicting incoming and partition batch sizes into BatchSizePredictorImpl. In unit tests various corner cases are tested by providing mock implementation of BatchSizePredictorImpl.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 1]
```diff
@@ -381,16 +409,14 @@ public HashJoinMemoryCalculator getCalculatorImpl() {
...
+    if (!prefetchedBuild) {
       // If we didn't retrieve our first data hold batch, we need to do it now.
-      prefetched = true;
-      prefetchFirstBatchFromBothSides();
+      prefetchedBuild = true;
+      prefetchFirstBuildBatch();
 
       // Handle emitting the correct outcome for termination conditions
-      // Use the state set by prefetchFirstBatchFromBothSides to emit the correct termination outcome.
+      // Use the state set by prefetchFirstBuildBatch to emit the correct termination outcome.
```
Comment (Ben-Zvi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: Code cleaning: The check below (switch statement) can be done as part of the identical check after executeBuildPhase() (which is skipped anyway if STOP).  Also maybe move the "wasKilled" check first thing in innerNext().

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 2]
```diff
@@ -248,32 +254,54 @@ protected void buildSchema() throws SchemaChangeException {
...
       // Got our first batche(s)
       state = BatchState.FIRST;
+    }
+  }
+
+  /**
+   *
+   * @return True terminate. False continue.
+   */
+  private boolean prefetchFirstProbeBatch() {
```
[the remark this answers is not in the document]
Comment (ilooner) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: Refactored this code.

**Comment by ilooner** [comment]
@Ben-Zvi I've refactored the code, should be much cleaner now. Please take another look.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 3]
```diff
@@ -248,95 +257,134 @@ protected void buildSchema() throws SchemaChangeException {
     }
   }
```
Comment (Ben-Zvi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: It would be nicer if *buildSchema()* was calling the "standard" *prefetchFirstBatchFromBothSides()* like the other binary operators (sans MJ). Instead of having specialized different code, and modifying the RecordBatch just for the hash join.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java [hunk 4]
```diff
@@ -248,95 +257,134 @@ protected void buildSchema() throws SchemaChangeException {
...
+      });
+
+    // Left and right sides must return a valid response and both sides cannot be NONE.
+    return (!leftUpstream.isError() && !rightUpstream.isError()) &&
+      (leftUpstream != IterOutcome.NONE && rightUpstream != IterOutcome.NONE);
+  }
+
+  private IterOutcome sniffNewSchema(final int index,
+                                     final RecordBatch batch,
+                                     final Runnable schemaSetter) {
```
Comment (Ben-Zvi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java]: The use of lambda function as a parameter just makes this code harder to read, and adds no real value.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinMemoryCalculatorImpl.java [hunk 5]
```diff
@@ -365,17 +299,20 @@ protected void initialize(boolean autoTune,
                               double loadFactor) {
       Preconditions.checkState(!firstInitialized);
       Preconditions.checkArgument(initialPartitions >= 1);
+      // If we had probe data before there should still be probe data now.
+      // If we didn't have probe data before we could get some new data now.
+      Preconditions.checkState(probeSizePredictor.hasData() && !probeEmpty || !probeSizePredictor.hasData());
```
Comment (Ben-Zvi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinMemoryCalculatorImpl.java]: A simpler rewrite of the condition:  **! (probeEmpty && ! probeSizePredictor.hasData())**

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/BatchSizePredictor.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+   * @return Gets the number of records computed in the call to {@link #updateStats()}. Returns 0 if {@link #hasData()} is false.
+   * @throws IllegalStateException if {@link #updateStats()} was never called.
+   */
+  int getNumRecords();
+
+  /**
+   * True if the input batch had records in the last call to {@link #updateStats()}. False otherwise.
+   * @return True if the input batch had records in the last call to {@link #updateStats()}. False otherwise.
+   */
+  boolean hasData();
```
[the remark this answers is not in the document]
Comment (ilooner) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/BatchSizePredictor.java]: Renamed to hadDataLastTime

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinMemoryCalculatorImpl.java (hunk 5 above)
Comment (ilooner) [reply exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinMemoryCalculatorImpl.java]: Done

**Comment by ilooner** [comment]
@Ben-Zvi Thanks for the review and suggestions. The code is much cleaner now.

**Comment by Ben-Zvi** [comment]
OK - tag it ready-to-commit ...
